### DISCUSSION TITLE
Oqtane.Upgrade

### DISCUSSION THREAD

**Discussion by TonyValenti** [oqtane/oqtane.framework#833] [General]
I'd like to clean up this code a bit.
I'd like to:
1. Factor out constants
2.  Change LangVersion to latest
3.  Nullable-enable it.
4.  Fix all CA warnings/messages.
5.  Async-ify it and specifically change Thread.Sleep to an await call.
6. [BREAKING CHANGE] Implement Mono.Options for command line argument parsing.

@sbwalker Which of the above are approved?

**Comment by sbwalker** [comment]
I would consider approving these changes as long as you are able to provide a detailed description of how this Upgrade project is used in the core framework and how you plan to test these changes. I am making this request because this is one of the more complex scenarios to understand ( ie. the run-time characteristics of .NET Core are quite different from the .NET Framework ). This is a case where the code refactoring is 10% of the effort and the other 90% will be spent verifying that it works as expected.

In regards to using Mono.Options - if this an external package/dependency then please refer to the "Managed Dependencies" section in the Philosophy (  https://www.oqtane.org/Resources/Blog/PostId/538/oqtane-philosophy ). The goal is to minimize external dependencies - so unless there is no other way to achieve a required goal, we should avoid it. In this particular case since the code already works in its current form it does not make sense to add a dependency - especially one that introduces a breaking change.

**Comment by TonyValenti** [comment]
@sbwalker I'm stepping through some of this code and I noticed something.

https://github.com/oqtane/oqtane.framework/blob/master/Oqtane.Upgrade/Program.cs#L26

Is that folder supposed to exist?  It doesn't seem to anywhere I look.

**Reply by sbwalker** [reply]
The "Framework" folder should be created automatically if you follow the intended upgrade flow. The Upgrade project is just one component that is used as part of an entire upgrade flow - the goal being to provide a streamlined framework upgrade experience ( ie. Wordpress ). To start the process you need to login to your Oqtane instance as the host user and go to the Admin Dashboard and select System Update. It will allow you to upload a framework package or download a framework package from Nuget ( as long as there is a newer version available ). Understanding the intended purpose and expected flow is critical to being able to verify any changes or refactoring to this area. This is why I made the request above for a defined testing approach. 

Please understand that it is not my intent to treat you differently than anyone else.... this scenario tends to happen a lot in open source projects... developers download the source code and want to refactor the logic based on their personal preference or tool recommendations, but do not understand the purpose of the code, how it works, or how to test it. They expect that some other member of the community will validate their changes - which usually ends up happening in production. So after 20 years of open source experience I am very wary of these "drive by" types of contributions, and in the areas of the code which I know are more complicated, I try to hold people to a higher standard.

**Reply by TonyValenti** [reply]
You and I are definitely on the same page.
